lib/vector-common/src/request_metadata.rs

```diff
@@ -1,16 +1,180 @@
+    /// Creates a new Tagged variant for when we do not need to track events by
+    /// tags.
+    #[must_use]
+    pub fn new_untagged() -> Self {
+        Self::Untagged {
+            size: CountByteSize(0, JsonSize::zero()),
+        }
+    }
+
+    /// Returns a `HashMap` of tags => event counts for when we are tracking by tags.
+    /// Returns `None` if we are not tracking by tags.
+    #[must_use]
+    pub fn sizes(&self) -> Option<&HashMap<EventCountTags, CountByteSize>> {
+        match self {
+            RequestCountByteSize::Tagged { sizes } => Some(sizes),
+            RequestCountByteSize::Untagged { .. } => None,
+        }
+    }
+
+    /// Returns a single count for when we are not tracking by tags.
+    #[must_use]
+    pub fn size(&self) -> Option<CountByteSize> {
+        match self {
+            RequestCountByteSize::Tagged { .. } => None,
+            RequestCountByteSize::Untagged { size } => Some(*size),
+        }
+    }
+
+    /// Adds the given estimated json size of the event to current count.
+    pub fn add_event<E>(&mut self, event: &E, json_size: JsonSize)
+    where
+        E: GetEventCountTags,
+    {
+        match self {
+            RequestCountByteSize::Tagged { sizes } => {
+                let size = CountByteSize(1, json_size);
+                let tags = event.get_tags();
+
+                match sizes.get_mut(&tags) {
+                    Some(current) => {
+                        *current += size;
+                    }
+                    None => {
+                        sizes.insert(tags, size);
+                    }
+                }
+            }
+            RequestCountByteSize::Untagged { size } => {
+                *size += CountByteSize(1, json_size);
+            }
+        }
+    }
+}
+
+impl From<CountByteSize> for RequestCountByteSize {
+    fn from(value: CountByteSize) -> Self {
+        Self::Untagged { size: value }
+    }
+}
+
+impl<'a> Add<&'a RequestCountByteSize> for RequestCountByteSize {
+    type Output = RequestCountByteSize;
+
+    fn add(self, other: &'a Self::Output) -> Self::Output {
+        match (self, other) {
+            (
+                RequestCountByteSize::Tagged { sizes: mut us },
+                RequestCountByteSize::Tagged { sizes: them },
+            ) => {
+                for (key, value) in them {
+                    match us.get_mut(key) {
+                        Some(size) => *size += *value,
+                        None => {
+                            us.insert(key.clone(), *value);
+                        }
+                    }
+                }
+
+                Self::Tagged { sizes: us }
+            }
+
+            (
+                RequestCountByteSize::Untagged { size: us },
+                RequestCountByteSize::Untagged { size: them },
+            ) => RequestCountByteSize::Untagged { size: us + *them },
+
+            // The following two scenarios shouldn't really occur in practice, but are provided for completeness.
```

Comment:
Whether to use `Tagged` or `Untagged` is determined by a global configuration option, so the whole system should settle on one variation. It would only be a bug if they were both used together. I'm now wondering if I should panic instead?

Reply:
If those states are truly unreachable, I think we should avoid putting code there and instead leave them `unreachable!("why")`.

Reply:
Hmm, unless… could the state change on reload to make it reachable?